File: tests/zenko_tests/e2e_config/locations.py
#!/usr/bin/env python

from boto3 import Session
import logging
import os

_log = logging.getLogger("end2end configuration")

IAM_ENDPOINT = os.getenv("IAM_ENDPOINT", "http://iam.zenko.local")
CRR_ROLE_NAME = os.getenv("CRR_ROLE_NAME", "crr-role")

def _setup_crr_iam_resources(account_creds):
    """
    Sets up a CRR site by creating the user, role and policy.
    :param account_creds: credentials of the crr site account
    :return: accessKey and secretKey of the created crr user
    """
    try:
        iam_client = Session(
            aws_access_key_id=account_creds["AccessKeyId"],
            aws_secret_access_key=account_creds["SecretAccessKey"],
            aws_session_token=account_creds["SessionToken"],
        ).client('iam', endpoint_url=IAM_ENDPOINT, region_name='us-east-1')

        user = iam_client.create_user(UserName="crr-user")
        credentials = iam_client.create_access_key(UserName="crr-user")

        iam_client.create_role(
            RoleName=CRR_ROLE_NAME,
            AssumeRolePolicyDocument='''{
                "Version": "2012-10-17",
                "Statement": [
                    {
                        "Effect": "Allow",
                        "Principal": {
                            "AWS": "''' + user["User"]["Arn"] + '''"
                        },
                        "Action": "sts:AssumeRole"
                    }
                ]
            }'''
        )

        policy = iam_client.create_policy(
            PolicyName="crr-policy",
            PolicyDocument='''{
                "Version": "2012-10-17",
                "Statement": [
                    {
                        "Effect": "Allow",
                        "Action": "s3:ReplicateObject",
                        "Resource": "arn:aws:s3:::*/*"
                    }
                ]
            }'''
        )

        iam_client.attach_role_policy(
            RoleName=CRR_ROLE_NAME,
            PolicyArn=policy["Policy"]["Arn"]
        )

        return {
            "accessKey": credentials["AccessKey"]["AccessKeyId"],
            "secretKey": credentials["AccessKey"]["SecretAccessKey"],
        }
    except Exception as e:
        raise Exception("Failed to setup CRR site: %s" % e)

def create_location(client, uuid, location, accounts_creds):
    """
    Creates a location
    :param client: swagger client
    :param uuid: zenko instance uuid
    :param location: location details
    :param account_credentials: credentials of the accounts created
    """
    
    ENABLE_RING_TESTS = os.environ['ENABLE_RING_TESTS']
    if ENABLE_RING_TESTS == "false" and location["locationType"] == "location-scality-ring-s3-v1":
        return

    DEPLOY_CRR_LOCATIONS = os.getenv('DEPLOY_CRR_LOCATIONS', 'true')
    if location["locationType"] == "location-scality-crr-v1":
        if DEPLOY_CRR_LOCATIONS == "false":
            return

        account_name = os.environ['CRR_SOURCE_ACCOUNT_NAME']
        if location["name"] == os.environ['CRR_DESTINATION_LOCATION_NAME']:
            account_name = os.environ['CRR_DESTINATION_ACCOUNT_NAME']

        user_creds = _setup_crr_iam_resources(accounts_creds[account_name])
        location["details"]["accessKey"] = user_creds["accessKey"]
        location["details"]["secretKey"] = user_creds["secretKey"]

    try:
        Location_V1 = client.get_model('location-v1')
        if "bootstrapList" not in location["details"]:
            location["details"]["bootstrapList"] = []
        loc = Location_V1(name=location["name"],
                            locationType=location["locationType"],
                            details=location["details"])

        res = (
            client.ui_facing
            .createConfigurationOverlayLocation(location=loc, uuid=uuid)
            .response()
            .result
        )

        _log.info("location %s created", location["name"])
    except Exception as e:
        raise Exception(
            "Failed to create location '%s': %s" % (location["name"], e))

File: tests/functional/testResourcesSetup/setup/locations.ts
import {
    IAMClient,
    CreateUserCommand,
    CreateAccessKeyCommand,
    CreateRoleCommand,
    CreatePolicyCommand,
    AttachRolePolicyCommand,
} from '@aws-sdk/client-iam';
import { PensieveClient } from './clients/pensieveClient';
import { LocationConfig, Env } from '../config';
import { AccountCredentials } from './clients/k8s';

interface CRRUserCredentials {
    accessKey: string;
    secretKey: string;
}

async function setupCRRIamResources(
    accountCreds: AccountCredentials,
    env: Env,
): Promise<CRRUserCredentials> {
    const iamClient = new IAMClient({
        credentials: {
            accessKeyId: accountCreds.AccessKeyId,
            secretAccessKey: accountCreds.SecretAccessKey,
            sessionToken: accountCreds.SessionToken,
        },
        endpoint: env.IAM_ENDPOINT,
        region: 'us-east-1',
    });

    const userResult = await iamClient.send(new CreateUserCommand({ UserName: 'crr-user' }));
    const accessKeyResult = await iamClient.send(new CreateAccessKeyCommand({ UserName: 'crr-user' }));

    await iamClient.send(new CreateRoleCommand({
        RoleName: env.CRR_ROLE_NAME,
        AssumeRolePolicyDocument: JSON.stringify({
            Version: '2012-10-17',
            Statement: [{
                Effect: 'Allow',
                Principal: { AWS: userResult.User?.Arn },
                Action: 'sts:AssumeRole',
            }],
        }),
    }));

    const policyResult = await iamClient.send(new CreatePolicyCommand({
        PolicyName: 'crr-policy',
        PolicyDocument: JSON.stringify({
            Version: '2012-10-17',
            Statement: [{
                Effect: 'Allow',
                Action: 's3:ReplicateObject',
                Resource: 'arn:aws:s3:::*/*',
            }],
        }),
    }));

    await iamClient.send(new AttachRolePolicyCommand({
        RoleName: env.CRR_ROLE_NAME,
        PolicyArn: policyResult.Policy?.Arn,
    }));

    if (!accessKeyResult.AccessKey?.AccessKeyId || !accessKeyResult.AccessKey?.SecretAccessKey) {
        throw new Error('CreateAccessKey did not return credentials');
    }
    return {
        accessKey: accessKeyResult.AccessKey.AccessKeyId,
        secretKey: accessKeyResult.AccessKey.SecretAccessKey,
    };
}

export async function createLocations(
    pensieveClient: PensieveClient,
    env: Env,
    locations: LocationConfig[],
    accountsCreds: Record<string, AccountCredentials>,
): Promise<void> {
    for (const location of locations) {
        if (!location.name) {
            console.warn(`Skipping location with empty name (type: ${location.locationType})`);
            continue;
        }

        if (!env.ENABLE_RING_TESTS && location.locationType === 'location-scality-ring-s3-v1') {
            continue;
        }

        if (location.locationType === 'location-scality-crr-v1') {
            if (!env.DEPLOY_CRR_LOCATIONS) {
                continue;
            }
            if (!env.CRR_SOURCE_ACCOUNT_NAME || !env.CRR_DESTINATION_ACCOUNT_NAME ||
                !env.CRR_DESTINATION_LOCATION_NAME) {
                throw new Error(
                    'CRR_SOURCE_ACCOUNT_NAME, CRR_DESTINATION_ACCOUNT_NAME, and ' +
                    'CRR_DESTINATION_LOCATION_NAME are required when DEPLOY_CRR_LOCATIONS=true',
                );
            }
            const locationAccountMap: Record<string, string | undefined> = {
                [env.CRR_DESTINATION_LOCATION_NAME ?? '']: env.CRR_DESTINATION_ACCOUNT_NAME,
                [env.CRR_LOCATION_A_NAME ?? '']: env.CRR_ACCOUNT_A_NAME,
                [env.CRR_LOCATION_B_NAME ?? '']: env.CRR_ACCOUNT_B_NAME,
                [env.CRR_LOCATION_C_NAME ?? '']: env.CRR_ACCOUNT_C_NAME,
            };
            const accountName = locationAccountMap[location.name] ?? env.CRR_SOURCE_ACCOUNT_NAME;
            if (!accountName || !accountsCreds[accountName]) {
                throw new Error(
                    `No credentials found for CRR location "${location.name}" ` +
                    `(resolved account: "${accountName ?? 'undefined'}"). ` +
                    'Ensure the corresponding CRR_*_ACCOUNT_NAME env var and k8s secret are set.',
                );
            }
            const userCreds = await setupCRRIamResources(accountsCreds[accountName], env);
            location.details.accessKey = userCreds.accessKey;
            location.details.secretKey = userCreds.secretKey;
        }

        await pensieveClient.createLocation(env.UUID, {
            name: location.name,
            locationType: location.locationType,
            details: location.details as Record<string, unknown>,
        });
        console.log(`Created location: ${location.name}`);
    }
}
